fate/src/main/java/org/apache/accumulo/fate/util/LoggingRunnable.java

```diff
@@ -47,7 +47,13 @@ public void run() {
         System.err
             .println("ERROR " + new Date() + " Exception that failed to log : " + t.getMessage());
         t.printStackTrace();
+
+        // If we got a more serious failure when attempting to log the message,
+        // then throw that instead of the original exception.
+        if ((t2 instanceof Error) && !(t instanceof Error))
```

Comment:
Did you consider doing this before the System.err.println?

Reply:
> Did you consider doing this before the System.err.println?

@keith-turner I hadn't considered that. Is your thinking that if the log call fails with an Error, then the likelihood of the println (or `new Date()`) calls failing too is higher? If those fail, then we'd end up propagating whatever exception they generated instead of either the original Throwable or the one that was thrown by the log call.

Reply:
>  Is your thinking that if the log call fails with an Error, then the likelihood of the println (or new Date()) calls failing too is higher?

Yeah I was wondering about that.  If there is an Error, it probably good to be cautious.  On the other hand attempting to get information out to a user is really important.  I like these changes as-is, but I was curious about others thoughts because I was a bit uncertain about it.

> If those fail, then we'd end up propagating whatever exception they generated instead of either the original Throwable or the one that was thrown by the log call.

That seems ok to me.  It may not be an error.

Reply:
> Yeah I was wondering about that. If there is an Error, it probably good to be cautious. On the other hand attempting to get information out to a user is really important. I like these changes as-is, but I was curious about others thoughts because I was a bit uncertain about it.

Agreed about being cautious and trying to get information to the user. Would it make sense to wrap the println calls in a try-catch block? If they throw an exception, then re-throw it if it's an Error and neither the original exception nor the exception thrown from the log call is an Error. The run method would look something like the following. Or is that overkill?
```java
  @Override
  public void run() {
    try {
      runnable.run();
    } catch (Throwable t) {
      boolean runnableError = (t instanceof Error);
      try {
        log.error("Thread \"{}\" died {}", Thread.currentThread().getName(), t.getMessage(), t);
      } catch (Throwable t2) {
        boolean logError = (t2 instanceof Error);
        try {
          // maybe the logging system is screwed up OR there is a bug in the exception, like
          // t.getMessage() throws a NPE
          System.err.println(
                  "ERROR " + new Date() + " Failed to log message about thread death " + t2.getMessage());
          t2.printStackTrace();

          // try to print original exception
          System.err
                  .println("ERROR " + new Date() + " Exception that failed to log : " + t.getMessage());
          t.printStackTrace();
        } catch (Throwable t3) {
          // If printing to System.err didn't work then don't try to log that exception but do re-throw it if
          // it's the most serious failure we've seen so far.
          boolean printError = (t3 instanceof Error);
          if (printError && !(logError || runnableError))
            throw t3;
        }

        // If we got a more serious failure when attempting to log the message,
        // then throw that instead of the original exception.
        if (logError && !runnableError)
          throw t2;
      }
      throw t;
    }
  }
```

Reply:
> Would it make sense to wrap the println calls in a try-catch block?

That seems reasonable.

> Or is that overkill?

It is unsightly code, not sure how to improve upon that though.  I like what it's trying to do, try as hard as possible to get information to a user and propagate errors.

Reply:
Ok, sounds good. I'm pushing that ugly code, and then I'll give it a bit in case anyone else has comments.

Reply:
I'm wondering if instead of 1 catch clause, you split it into two catch clauses. The first being `catch(Exception)`, which is logged and thrown, and the second being `catch(Throwable)` which calls `e.printStackTrace()` and then throws the Error. In the `Exception` case logging should not be an issue. However, in the `OutOfMemoryError` case, logging will be an issue.

Reply:
Logging may also be an issue in a StackOverflowError case too.

Reply:
You may want to mimick what I did [here](https://github.com/apache/accumulo/pull/1818/files#diff-f7183fa4f5966fa166de65c6e0bbf1e03368cf4eda9f1e683281e528c9406627) sans the server halt.

Reply:
I thought about having multiple catch blocks initially, but was feeling like I was going to duplicate all of the code. I guess it depends on how pedantic we want to be here. If I split it into 2 catches for Exception and Throwable, then I'd still need to wrap the log in the Exception catch clause in a try-catch to handle cases such as problems with the logging system or the getMessage on the exception itself, and then the code is effectively duplicated in both catch blocks.

Based on reading the code from your PR, I think maybe you're suggesting to log (but not try to fall back to printing to stderr) in the Exception catch block, and keep the current behavior for the Throwable catch block? Since this code isn't in an uncaught exception handler, I'd still need to catch Exception and ignore it in the first Exception catch block so that an Exception from the log call doesn't mask the original Exception produced by `runnable.run()`. I have no problem changing it if you think that's better. I'm not seeing the benefit in functionality or readability, but maybe I misunderstood what you are suggesting?